internal/cmd/commands/dev/dev.go

```diff
@@ -824,73 +823,84 @@ func (c *Command) Run(args []string) int {
+						c.worker.WorkerAuthStorage.Cleanup()
+					}
+					if err := c.worker.GracefulShutdown(); err != nil {
+						c.UI.Error(fmt.Errorf("Error shutting down worker gracefully: %w", err).Error())
+					}
+					if err := c.worker.Shutdown(); err != nil {
+						c.UI.Error(fmt.Errorf("Error shutting down worker: %w", err).Error())
+					}
+				}
 
-				case <-c.ServerSideShutdownCh:
-					// Drain connections in case this is hit more than once
+				if err := c.controller.Shutdown(); err != nil {
+					c.UI.Error(fmt.Errorf("Error shutting down controller: %w", err).Error())
+				}
 
-				case <-abortForceShutdownCh:
-					// No-op, we just use this to shut down the goroutine
-					return
+				err := c.opsServer.Shutdown()
+				if err != nil {
+					c.UI.Error(fmt.Errorf("Failed to shutdown ops listeners: %w", err).Error())
 				}
-			}
-		}()
+				shutdownDone = true
+			}()
+		}
+
+		if sync.LoadUint32(&shutdownTriggerCount) == 2 && !c.flagControllerOnly {
+			go func() {
+				if !c.flagControllerOnly {
+					if err := c.worker.Shutdown(); err != nil {
+						c.UI.Error(fmt.Errorf("Error shutting down worker: %w", err).Error())
+					}
+					shutdownDone = true
+				} else {
+					c.UI.Error("Forcing shutdown")
+					os.Exit(base.CommandUserError)
+				}
+			}()
+		}
 
-		shutdownTriggered = true
+		if sync.LoadUint32(&shutdownTriggerCount) > 2 {
+			go func() {
+				c.UI.Error("Forcing shutdown")
+				os.Exit(base.CommandUserError)
+			}()
+		}
 	}
 
-	for !shutdownTriggered && !errorEncountered.Load() {
+	for !errorEncountered.Load() && !shutdownDone {
 		select {
 		case <-c.ServerSideShutdownCh:
 			c.UI.Output("==> Boundary dev environment self-terminating")
+			sync.AddUint32(&shutdownTriggerCount, 1)
 			runShutdownLogic()
 
 		case <-c.ShutdownCh:
 			c.UI.Output("==> Boundary dev environment shutdown triggered, interrupt again to force")
+			sync.AddUint32(&shutdownTriggerCount, 1)
 			runShutdownLogic()
 
 		case <-c.SigUSR2Ch:
 			buf := make([]byte, 32*1024*1024)
 			n := runtime.Stack(buf[:], true)
 			event.WriteSysEvent(context.TODO(), op, "goroutine trace", "stack", string(buf[:n]))
-		}
-
-		if shutdownTriggered {
-			if c.Config.Controller != nil {
-				c.opsServer.WaitIfHealthExists(c.Config.Controller.GracefulShutdownWaitDuration, c.UI)
-			}
-
-			if !c.flagControllerOnly {
-				if !c.flagWorkerAuthStorageSkipCleanup && c.worker.WorkerAuthStorage != nil {
-					c.worker.WorkerAuthStorage.Cleanup()
-				}
-				if err := c.worker.Shutdown(); err != nil {
-					c.UI.Error(fmt.Errorf("Error shutting down worker: %w", err).Error())
-				}
-			}
-
-			if err := c.controller.Shutdown(); err != nil {
-				c.UI.Error(fmt.Errorf("Error shutting down controller: %w", err).Error())
-			}
-
-			err := c.opsServer.Shutdown()
-			if err != nil {
-				c.UI.Error(fmt.Errorf("Failed to shutdown ops listeners: %w", err).Error())
+		default:
+			if shutdownDone {
```

Comment:
This will cause very rapid spinning in the for loop because it will keep selecting this case over and over until one of the others is true. What I think you should do instead of `default` is do something like `time.After(10*time.Millisecond)` so that we don't end up creating a spin loop but also can react quickly when it's done.